californium-core/src/main/java/org/eclipse/californium/core/network/InMemoryMessageIdProvider.java

```diff
@@ -51,17 +54,26 @@ public InMemoryMessageIdProvider(final NetworkConfig config) {
 	}
 
 	@Override
-	public synchronized int getNextMessageId(final InetSocketAddress destination) {
+	public int getNextMessageId(final InetSocketAddress destination) {
+		MessageIdTracker tracker = getTracker(destination);
+		if (tracker == null) {
+			// we have reached the maximum number of active peers
+			// TODO: throw an exception?
+			return -1;
+		} else {
+			return tracker.getNextMessageId();
+		}
+	}
+
+	private synchronized MessageIdTracker getTracker(final InetSocketAddress destination) {
 		MessageIdTracker tracker = trackers.get(destination);
 		if (tracker == null) {
 			// create new tracker for destination lazily
 			tracker = new MessageIdTracker(config);
 			if (!trackers.put(destination, tracker)) {
-				// we have reached the maximum number of active peers
-				// TODO: throw an exception?
-				return -1;
+				return null;
```

Comment:
since we're synchronizing this call we could also check capacity first, instead of instantiating the `MessageIdTracker` and throwing it away if the map is already full?